projects/hathor-wallet-mobile/0003-token-swap/token-swap.md

```diff
@@ -0,0 +1,282 @@
+  - To configure the bucket to serve static content
+- `aws_s3_object`
+  - To upload the files
+
+### Dozer token swap protocol methods
+
+##### View methods
+
+These methods do not require a transaction to be sent, but the code is executed on the fullnode each time these are calculated.
+
+`find_best_swap_path(self, amount_in: Amount, token_in: TokenUid, token_out: TokenUid, max_hops: int) -> SwapPathInfo`
+- `amount_in`: The amount of input tokens
+- `token_in`: The input token uid
+- `token_out`: The output token uid
+- `max_hops`: Maximum number of hops (1-3, but algorithm handles any number)
+
+`find_best_swap_path_exact_output(self, amount_out: Amount, token_in: TokenUid, token_out: TokenUid, max_hops: int) -> SwapPathExactOutputInfo`
+- `amount_out`: The desired output amount
+- `token_in`: The input token
+- `token_out`: The output token
+- `max_hops`: Maximum number of hops (1-3, but algorithm handles any number)
+
+##### Public methods
+
+These methods require a transaction to be sent, the swap methods also require a deposit and withdraw action to swap the expected values of the tokens.
+ 
+`swap_exact_tokens_for_tokens(self, ctx: Context, fee: Amount) -> SwapResult`
+- `ctx`: The transaction context
+- `fee`: Fee for the pool (default: use default_fee)
+- The method also expects a deposit and withdraw actions with the swapped tokens.
+
+`swap_exact_tokens_for_tokens_through_path(self, ctx: Context, path_str: str) -> SwapResult`
+- `ctx`: The transaction context
+- `path_str`: Comma-separated string of pool keys to traverse
+- The method also expects a deposit and withdraw actions with the swapped tokens.
+
+`swap_tokens_for_exact_tokens(self, ctx: Context, fee: Amount) -> SwapResult`
+- `ctx`: The transaction context
+- `fee`: Fee for the pool (default: use default_fee)
+- The method also expects a deposit and withdraw actions with the swapped tokens.
+
+`swap_tokens_for_exact_tokens_through_path(self, ctx: Context, path_str: str) -> SwapResult`
+- `ctx`: The transaction context
+- `path_str`: Comma-separated string of pool keys to traverse
+- The method also expects a deposit and withdraw actions with the swapped tokens.
+
+##### Type definitions
+
+
+```py
+class SwapPathInfo(NamedTuple):
+    """Information about the best swap path between tokens."""
+
+    path: str
+    amounts: list[Amount]
+    amount_out: Amount
+    price_impact: Amount
+
+
+class SwapPathExactOutputInfo(NamedTuple):
+    """Information about the best swap path for exact output swaps."""
+
+    path: str
+    amounts: list[Amount]
+    amount_in: Amount
+    price_impact: Amount
+
+class SwapResult(NamedTuple):
+    """Result for an executed swap with the details of the execution.
+
+    Notice that the results are presented for tokens in and tokens out.
+    So one must check which one is Token A and which one is Token B."""
+
+    amount_in: Amount
+    slippage_in: Amount
+    token_in: TokenUid
+    amount_out: Amount
+    token_out: TokenUid
+```
+

```

Comment:
I would also add the feature toggle to turn on/off this feature and how the screen will behave

Reply:
Done